DOC: Update randn() to use rng.standard_normal()

bmwoodruff opened the pull request:
Updates all references to `randn` to use `standard_normal` from Generator. I left the matlib versions alone.
In a few spots I added `seed=123`.
I left the `# may vary` tag on those examples.
This tag could be removed with the guaranteed same output now. I had to reformat a few chunks of longer code and output.

[skip actions] [skip azp] [skip cirrus]

This PR stems from a [Slack discussion](https://numpy-team.slack.com/archives/CM8C6PE67/p1715718898610159).  

I tested removing the `# may vary` tag with `python tools/refguide_check.py --doctests`, and things worked fine.  I obtained the same outputs on a dev build on Linux, and an old pip installed numpy on Windows. Should I remove the `# may very` tag?  

I'm planning on replacing other instances of `random` to use the updated modules. This was just replacing `randn`. I'll incorporate feedback from this PR in those.

bmwoodruff committed:
DOC: Update randn() to use rng.standard_normal()

Updates all references to `randn` to use `standard_normal` from Generator.
I left the matlib versions alone.
In a few spots I added `seed=123`.
I left the `# may vary` tag on those examples.
This tag could be removed with the guaranteed same output now.
I had to reformat a few chunks of longer code and output.

[skip actions] [skip azp] [skip cirrus]

mattip commented:
> Should I remove the # may very tag?

I think so, although that could be a separate task, and might wait for a refactoring of doctests, see #15845, #26604 and others.

mattip commented on numpy/linalg/_linalg.py:
```diff
@@ -1703,8 +1707,13 @@ def svd(a, full_matrices=True, compute_uv=True, hermitian=False):
+    >>> a = a_re + 1j*a_im
+    >>> b_re = rng.standard_normal(size=(2, 7, 8, 3))
+    >>> b_im = rng.standard_normal(size=(2, 7, 8, 3))
+    >>> b = b_re + 1j*b_im
```
This is a bit awkward, but I am not sure how to more cleanly create `a` and `b`

bmwoodruff replied:
I agree completely.  I tried multiple ways to rewrite this part, and all were worse that the above. I'll give `norm` a try and see if that improves it.

bmwoodruff replied:
For anyone reading this in the future, note that the new version is `normal` rather than `norm`.  Too many functions too keep track of.

bmwoodruff replied:
The shorter function name kept the line length to 77, so it's back to the shorter version.

mattip commented on numpy/polynomial/hermite.py:
```diff
@@ -1460,10 +1460,11 @@ def hermfit(x, y, deg, rcond=None, full=False, w=None):
     >>> x = np.linspace(-10, 10)
-    >>> err = np.random.randn(len(x))/10
+    >>> rng = np.random.default_rng(seed=123)
+    >>> err = rng.standard_normal(size=(len(x)))/10
```
```suggestion
    >>> err = rng.norm(scale=1./10., size=len(x))
```

bmwoodruff replied:
I like the change. The extra parenthesis were awkward. I tried to preserve as much of the original example as I could, rather than thinking about using a different function (which here makes it better). I'll make the change here, and below, and I think there is one more as well.

mattip commented on numpy/polynomial/hermite_e.py:
```diff
@@ -1387,11 +1387,11 @@ def hermefit(x, y, deg, rcond=None, full=False, w=None):
-    >>> np.random.seed(123)
-    >>> err = np.random.randn(len(x))/10
+    >>> rng = np.random.default_rng(seed=123)
+    >>> err = rng.standard_normal(size=(len(x)))/10
```
use norm(scale..., remove excess parentheses

mattip reviewed (commented):
Nice. A few suggestions, mainly around using `norm(scale=...` instead.

Edit: I did not comment on every use of `scale`.

bmwoodruff commented:
> Curious why you chose `standard_normal` instead of `norm`, which by default is equivalent and shorter. I am not sure which use we should encourage.

I followed the suggested new use directly, without looking at the bigger picture and alternate options:
https://github.com/numpy/numpy/blob/c41621b3bb7c5e387049bab25794090635a2faa4/numpy/random/mtrand.pyx#L1256-L1260

I think the suggested use above is correct, for someone porting code from Matlab who wants to guarantee that it's the standard normal distribution. The examples I changed in the PR are not about porting code from Matlab, so I think using the shorter `normal` option will be much better for all involved.  

I vote for encouraging the shorter use.  I'll simplify the code in the PR.  Thanks for the suggestions.

bmwoodruff commented on numpy/polynomial/polynomial.py:
```diff
@@ -1441,27 +1441,31 @@ def polyfit(x, y, deg, rcond=None, full=False, w=None):
     >>> stats # note the minuscule SSR
-    [array([  7.46346754e-31]), 4, array([ 1.38446749,  1.32119158, # may vary
-               0.50443316,  0.28853036]), 1.1324274851176597e-014]
+    [array([8.79579319e-31]),                                 # may vary
+     4,                                                       # may vary
+     array([1.38446749, 1.32119158, 0.50443316, 0.28853036]), # may vary
+     1.1324274851176597e-14]                                  # may vary
```
I would love some feedback on this particular change. Rather than try to cut a really long output at one spot, I used the Jupyter formatting.  Good/bad?

mattip replied:
As long as doctests are OK with it, I think this is better.

bmwoodruff replied:
Looks like the tests passed.  

Though something is off with the docstring tester, as I purposefully entered some arbitrary changes to make it fail, yet the tests passed. I'll be exploring that more and open an issue if/when I have something reproducible with a bit more to go on.

bmwoodruff committed:
DOC: Update randn() to use rng.normal()

Updates all references to `randn` to use `normal` from Generator.
I left the matlib versions alone.
In a few spots I added `seed=123`.
I removed the `# may vary` tag from all examples.
I reformatted some longer outputs to use Jupyter style output.

[skip actions] [skip azp] [skip cirrus]

rkern commented on numpy/polynomial/hermite.py:
```diff
@@ -1460,10 +1460,11 @@ def hermfit(x, y, deg, rcond=None, full=False, w=None):
     >>> y = hermval(x, [1, 2, 3]) + err
     >>> hermfit(x, y, 2)
-    array([1.0218, 1.9986, 2.9999]) # may vary
+    array([1.02294967, 2.00016403, 2.99994614])
```
Please omit the `seed` and continue to use the `# may vary`. In our [`np.random` documentation](https://numpy.org/doc/stable/reference/random/index.html) we recommend using large seeds and avoid small common seeds like `123`, and we want to demonstrate this practice in all of our examples. Since none of these are actually demonstrating seeding, we prefer to just use `np.random.default_rng()` with no seed.

In the scipy documentation, our documentation generation and doctesting framework will replace an argumentless `np.random.default_rng()` with one with a common large seed to get reproducibility of the example. We might import that practice here at some point, but for now `# may vary` is fine.

bmwoodruff replied:
Done. I like the future plan of using a common large seed. I knew about the recommendation of using large seeds, just wasn't sure removing it from a current example was wanted.  I'll strip small seeds from future PRs as I go. Thanks for the feedback.

bmwoodruff committed:
DOC: Update randn() to use rng.normal()

Updates all references to `randn` to use `normal` from Generator.
I left the matlib versions alone.
I remove all references to `seed=123` and left  `# may vary`.
Longer outputs were reformatted to use Jupyter style output.

[skip actions] [skip azp] [skip cirrus]

mattip commented on numpy/polynomial/hermite.py:
```diff
@@ -1460,10 +1460,11 @@ def hermfit(x, y, deg, rcond=None, full=False, w=None):
     >>> x = np.linspace(-10, 10)
-    >>> err = np.random.randn(len(x))/10
+    >>> rng = np.random.default_rng()
+    >>> err = rng.normal(scale=1./10,size=len(x))
```
```suggestion
    >>> err = rng.normal(scale=1./10, size=len(x))
```

mattip commented on numpy/polynomial/hermite_e.py:
```diff
@@ -1387,11 +1387,11 @@ def hermefit(x, y, deg, rcond=None, full=False, w=None):
-    >>> np.random.seed(123)
-    >>> err = np.random.randn(len(x))/10
+    >>> rng = np.random.default_rng()
+    >>> err = rng.normal(scale=1./10,size=len(x))
```
```suggestion
    >>> err = rng.normal(scale=1./10, size=len(x))
```

mattip commented on numpy/polynomial/laguerre.py:
```diff
@@ -1441,10 +1441,11 @@ def lagfit(x, y, deg, rcond=None, full=False, w=None):
     >>> x = np.linspace(0, 10)
-    >>> err = np.random.randn(len(x))/10
+    >>> rng = np.random.default_rng()
+    >>> err = rng.normal(scale=1./10,size=len(x))
```
```suggestion
    >>> err = rng.normal(scale=1./10, size=len(x))
```

bmwoodruff replied:
Thank you for the white space reminder and for letting me fix it rather than just changing it and merging it in. The training is helpful.

mattip commented on numpy/polynomial/polynomial.py:
```diff
@@ -1441,27 +1441,31 @@ def polyfit(x, y, deg, rcond=None, full=False, w=None):
+    [array([48.312088]),                                        # may vary
+     4,
+     array([1.38446749, 1.32119158, 0.50443316, 0.28853036]),
+     1.1324274851176597e-14]
```
I am a bit surprised doctests accepts the formatting change. Is this doctest-ed?

bmwoodruff replied:
It was doctested.  I was a bit surprised too.  I think something is causing it to miss this function. It's on my to-do list to explore why it was skipped.

rkern replied:
Our doctest runner [normalizes whitespace by default](https://github.com/numpy/numpy/blob/main/tools/refguide_check.py#L803). IPython-pretty-printed lists are acceptable (see my own journey of discovery [on this other issue](https://github.com/numpy/numpy/pull/26538#discussion_r1616134523)).

mattip replied:
I think the `mayvary` directive is telling doctests to ignore everything. You can run this locally (after `pip install .`) with the verbose flag and then see the doctest output
```
python tools/refguide_check.py --doctests numpy.polynomial -vv > /tmp/output
```

Basically, anything there will be accepted.

rkern replied:
That too. :smile:

bmwoodruff replied:
@mattip , I checked it yesterday without the `# may vary` tag and things ran fine. Thanks for the link @rkern on your journey, I'll be following that today. 

The issue I found yesterday is that without the `# may vary` tag, If I changed the output (say a 2.999... to a 5.9999), then the test still passed on multiple of the functions in the `numpy polynomial` module.  I was able to break the doctest in other places, but not there.  I have to spend some time to make something reproducible, then I'll start a different issue that references this discussion (or fix the issue once I narrow it down).

bmwoodruff replied:
It was a `user error`. :)  No issue with the doc test.  I followed the workflow in `.circleci/config.yaml`, and found the missing bit.  You mentioned it above, namely `pip install .`

ev-br replied:
> the mayvary directive is telling doctests to ignore everything

As a belated comment, `# may vary` make the doctester ignore the output (https://github.com/scipy/scipy_doctest/blob/main/scipy_doctest/impl.py#L271)

The difference to `# doctest: +SKIP` is two-fold:
- `# doctest: +SKIP` should be added the example (the line starting with `>>>`, in the doctest parlance, it's `example.source`); `# may vary` goes with the output (in the doctest language, `want`)
- `# doctest: +SKIP` skips checking that the example is legal python; `# may vary` does not; so it's really "check that the example.source produces something, and don't bother checking this something"

bmwoodruff committed:
DOC: Update randn() to use rng.normal()

Updates all references to `randn` to use `normal` from Generator.
I left the matlib versions alone.
I removed all references to `seed=123` and left  `# may vary`.
Longer outputs were reformatted to use Jupyter style output.

[skip actions] [skip azp] [skip cirrus]

mattip commented:
Thanks @bmwoodruff
